feat: add azure openai support

thegovind opened the pull request:
## Summary

Adds optional Azure OpenAI support alongside the existing OpenAI flow. When `AZURE_OPENAI_ENDPOINT` (plus related env vars) is set, Codex will authenticate via Entra ID and call the Responses API against your Azure OpenAI deployments. Otherwise it falls back to the standard `OPENAI_API_KEY` path.

Closes #11 #174 

---

## What’s Changed

- **Model & Agent code**  
  - **`model‑utils.ts`**  
    - Detects `AZURE_OPENAI_ENDPOINT`; if present, uses `@azure/identity`’s `DefaultAzureCredential` and the SDK’s `AzureOpenAI` client to list deployments via Responses API.  
    - Falls back to `OPENAI_API_KEY` flow or default model list on errors.  
  - **`agent-loop.ts`**  
    - Updated reasoning initialization to work correctly when using Azure models.
- **Configuration**  
  - Exported four new env vars in `config.ts`:  
    - `AZURE_OPENAI_ENDPOINT`  
    - `AZURE_OPENAI_API_VERSION`  
    - `AZURE_OPENAI_DEPLOYMENT` (optional) 
    - `AZURE_OPENAI_API_KEY` (fallback) 
- **CLI error handling**  
  - `cli.tsx` now errors if neither `OPENAI_API_KEY` nor `AZURE_OPENAI_ENDPOINT` is configured.  
  - Enhanced error messaging to guide through `az login` and Azure env‑var setup.
- **README updates**  
  - Added a “Using Azure OpenAI” section under Configuration with bash export examples.  
  - Clarified that Entra ID auth (via Azure AD) is supported for Azure OpenAI.
- **Dependencies**  
  - Added `@azure/identity` and bumped `openai` to include `AzureOpenAI` support.

---

## How to Test

1.  **Deploy your model**
    - Go to [[Azure AI Foundry](https://ai.azure.com/)](https://ai.azure.com) and deploy either `o4‑mini` or `o3` to your Azure OpenAI resource.
2.  **Set up env vars**
    ```bash
    az login
    export AZURE_OPENAI_ENDPOINT="https://<your‑resource>.openai.azure.com"
    export AZURE_OPENAI_API_VERSION="2024-02-01"
    export AZURE_OPENAI_DEPLOYMENT="your-deployment-name" # optional
    export AZURE_OPENAI_API_KEY="your-api-key" # if not present, it'll use Entra ID
    ```
3.  **Get the Codex CLI by building from your PR branch**
    Clone your specific PR branch and build the CLI:

    ```bash
    # Clone your fork of the repository
    git clone https://github.com/thegovind/codex.git

    # Navigate into the cloned repository
    cd codex

    # Checkout your specific PR branch
    git checkout gk/azure-openai

    # Navigate to the CLI package directory
    cd codex-cli

    # Enable corepack
    corepack enable

    # Install dependencies and build
    pnpm install
    pnpm build

    # Get the usage and the options (optional)
    node ./dist/cli.js --help

    # After building, you can either:
    # - Run the locally-built CLI directly using `node ./dist/cli.js <args>`
    # - Link the command globally for convenience using `pnpm link` (then use `codex <args>`)
    ```
4.  **Invoke Codex**
    Run the locally built CLI using `node` from the `codex-cli` directory:

    ```bash
    node ./dist/cli.js "explain this function"
    ```
    *(Alternatively, if you ran `pnpm link` in the previous step, you can simply use `codex "explain this function"` from anywhere)*.

---

## Tests

- [x] Updated and ran all tests successfully 
- [x] ~~Update default API version to `2025-04-01-preview` once it’s available in Azure OpenAI~~
  *(Note: [Temporary Conditional logic](https://github.com/openai/codex/pull/92#issuecomment-2816919410) added for Azure OpenAI will be removed in another PR)*

---

## Checklist

- [x] Model listing and agent loop updated for Azure path  
- [x] Config exports for Azure env vars  
- [x] CLI guards/error messages for missing credentials  
- [x] README examples for Azure OpenAI  
- [x] Added `@azure/identity`, bumped `openai` dependency  
- [x] Write and update tests for Azure code path  

cc @theabhinavdas @nikodem-wrona for SDK review.

thegovind committed:
add azure openai support

theabhinavdas commented on codex-cli/package.json:
`@azure/identity` doesn't seem to be used. also, wouldn't you need `@azure/openai` for the diff on line 2 in `codex-cli/src/utils/model-utils.ts`?

thegovind replied:
@theabhinavdas 

I’ve refactored our Azure auth examples to standardize on Entra ID and removed the API‑key flow. We have to update [our official docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/supported-languages?tabs=dotnet-secure%2Capi-key%2Cpython-secure%2Ccommand&pivots=programming-language-javascript) on our end, but the ones on your official SDK repo are accurate - see [azure.md](https://github.com/openai/openai-node/blob/HEAD/[azure.md](https://github.com/openai/openai-node/blob/HEAD/azure.md))

theabhinavdas replied:
Nice! Thanks, @thegovind!

alioftech commented:
That was fast! Was looking for this. Thanks

thegovind committed:
only use entra id flow for AOAI

thegovind committed:
Merge branch 'main' into gk/azure-openai

thegovind committed:
add updated package-lock for ci build error

thegovind commented:
I have read the CLA Document and I hereby sign the CLA

thegovind committed:
fix tests and linter issues

thegovind committed:
Merge branch 'main' into gk/azure-openai

junseinagao commented:
@thegovind

Hi,

I tried running this PR, but Codex still throws the error “Missing OpenAI API key.” even though I’ve set the following environment variables: `AZURE_OPENAI_ENDPOINT`, `AZURE_OPENAI_API_VERSION`, and `AZURE_OPENAI_DEPLOYMENT`.

It looks like we might also need to update [cli.tsx](https://github.com/openai/codex/blob/2aa8ae4167a70840c4f237bf960fe608a78d97d5/codex-cli/src/cli.tsx#L209-L220).

thegovind committed:
update Azure OpenAI instructions for Entra ID

thegovind committed:
Merge branch 'main' into gk/azure-openai

theabhinavdas commented:
@thegovind Adding tests will also help.

thegovind committed:
updated cli instructions and fixed local build where zod was missing

thegovind commented:
> @thegovind
> 
> Hi,
> 
> I tried running this PR, but Codex still throws the error “Missing OpenAI API key.” even though I’ve set the following environment variables: `AZURE_OPENAI_ENDPOINT`, `AZURE_OPENAI_API_VERSION`, and `AZURE_OPENAI_DEPLOYMENT`.
> 
> It looks like we might also need to update [cli.tsx](https://github.com/openai/codex/blob/2aa8ae4167a70840c4f237bf960fe608a78d97d5/codex-cli/src/cli.tsx#L209-L220).

@junseinagao Thanks for testing. I’ve only added support for Entra ID, so you’ll need to run `az login` locally before running this and ensure the relevant Azure environment variables are set. I’ve also updated the CLI instructions as you suggested. Thank you!

adhishthite commented:
@thegovind 

Thanks for this, can we make it work with non-Entra ID Azure OpenAI as well, i.e.: Just the AZURE_OPENAI_KEY and stuff? Would be much more helpful!

Thanks!

adhishthite commented:
@thegovind 

https://github.com/openai/codex/issues/174

Since you're working on this, can you also take a look at this issue above - so that a TON of people will benefit? Much appreciated.

tibo-openai commented:
Would it be possible to build this in a way that doesn't depend on "@azure/identity"?

thegovind commented:
@adhishthite @tibo-openai @junseinagao @theabhinavdas @alioftech and others following this - Thanks for your patience. I’ll be pushing more updates soon, but we’ll need to wait for our Azure OpenAI Inference team to update the underlying APIs before everything works end-to-end.

👉 The latest API version from us needs to be on parity with OpenAI (imminent):  
[Azure OpenAI Inference Specs](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview)

If you try to use the most recent spec ([`2025-03-01-preview`](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2025-03-01-preview)), you’ll likely run into this:

![image](https://github.com/user-attachments/assets/2a49eb64-a6fc-4f82-805e-56e5299a4df8)

So in short: let’s not put the cart before the horse 😄 I’ll follow up with more concrete guidance once the API updates land this week or next.

pskoett commented:
This is great, win win

moosh3 commented:
Nice!

APiTJLillo commented:
Amazing, keeping an eye on this!

adamgoslin commented:
@thegovind Any timeline for the update to the Azure openai inference api? If the release is delayed, it might be wise to add the conditional for AOAI and get this merged. You can always come back and remove it.

thegovind committed:
Merge branch 'main' into gk/azure-openai

thegovind committed:
feat: update Azure OpenAI configuration and enhance model fetching

thegovind committed:
Merge branch 'main' into gk/azure-openai

thegovind committed:
update tests to incorproate azure openai

thegovind committed:
Merge remote-tracking branch 'origin/gk/azure-openai' into gk/azure-openai

thegovind committed:
Merge branch 'main' into gk/azure-openai

thegovind committed:
add aoai temp conditional for reasoning summary

thegovind committed:
Merge remote-tracking branch 'origin/gk/azure-openai' into gk/azure-openai

thegovind committed:
Merge branch 'main' into gk/azure-openai

thegovind commented:
> @thegovind Any timeline for the update to the Azure openai inference api? If the release is delayed, it might be wise to add the conditional for AOAI and get this merged. You can always come back and remove it.

Okay @adamgoslin , I added the conditional and also updated the tests. @theabhinavdas @nikodem-wrona - please review and merge. I'll remove the [conditional](https://github.com/openai/codex/pull/92/files#diff-b15957eac2720c3f1f55aa32f172cdd0ac6969caf4e7be87983df747a9f97083R523) once we have `2025-04-01-preview` inference API in our [data plane](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview) (current ETA is 4/22). It works fine with o3 and o4-mini in AOAI:
![image](https://github.com/user-attachments/assets/50b80830-8413-4490-829a-2796354dff1c)

![image](https://github.com/user-attachments/assets/d61e5f22-c099-475a-85cd-2bd21d7e1095)

thegovind committed:
Merge branch 'main' into gk/azure-openai

theabhinavdas commented on README.md:
````diff
@@ -312,7 +312,32 @@ You can also define custom instructions:
+To use Azure OpenAI with Codex, set the following environment variables:
+
+```bash
+# Set Azure OpenAI credentials
````
Lines 325 through 337 look repetitive? Lines 324 through 326 already have this info.

thegovind replied:
Removed the section and updated the references

bsormagec commented:
Why are you using this unconventional method? Why not implement API key-based authentication instead of Entra ID? Is there a specific reason behind this choice? 

`The `apiKey` and `azureADTokenProvider` arguments are mutually exclusive; only one can be passed at a time.`

I encounter this error when building and attempting to run with az login.

thegovind committed:
update readme to remove duplicate AOAI env vars

thegovind committed:
Merge remote-tracking branch 'origin/gk/azure-openai' into gk/azure-openai

thegovind committed:
Merge branch 'main' into gk/azure-openai

theabhinavdas commented on codex-cli/src/utils/config.ts:
```diff
@@ -36,6 +36,12 @@ export const OPENAI_TIMEOUT_MS =
+export const AZURE_OPENAI_API_VERSION =
+  process.env["AZURE_OPENAI_API_VERSION"] || "2025-03-01-preview";
+export const AZURE_OPENAI_DEPLOYMENT =
+  process.env["AZURE_OPENAI_DEPLOYMENT"] || "04-mini";
```
Assuming you mean `o4-mini` and not `04-mini`?

thegovind replied:
Good catch. Fixed it. OpenAI's naming conventions are confusing everyone :)

thegovind commented:
> Why are you using this unconventional method? Why not implement API key-based authentication instead of Entra ID? Is there a specific reason behind this choice?
> 
> `The `apiKey`and`azureADTokenProvider` arguments are mutually exclusive; only one can be passed at a time.`
> 
> I encounter this error when building and attempting to run with az login.

I'll work on the API Key support in another PR. This one already has too many changes. Meanwhile, could you please ensure the `OPENAI_API_KEY` environment variable is unset when you test with Entra ID authentication?

You can typically unset it for your current terminal session using:

* **Bash (Linux/macOS/WSL):**
    ```bash
    unset OPENAI_API_KEY
    ```
* **PowerShell (Windows/Cross-Platform):**
    ```powershell
    Remove-Item Env:OPENAI_API_KEY
    ```

Make sure you only have the necessary Azure-specific environment variables configured for token authentication, such as:

* `AZURE_OPENAI_ENDPOINT`: Your specific Azure OpenAI resource endpoint URL. (Required)
* `AZURE_OPENAI_API_VERSION`: Optional. 
* `AZURE_OPENAI_DEPLOYMENT`: Optional. (Defaults to `04-mini` if not set.)

As for *why* we lean towards Entra ID (Azure AD) token authentication:

1.  **Enhanced Security:** Token-based authentication is generally more secure. Tokens are typically short-lived and managed via an identity provider, reducing the risks associated with static, long-lived API keys which, if compromised, provide persistent access until revoked.
2.  **Enterprise Requirements:** Many organizations, especially larger enterprises using Azure, have security policies that mandate or strongly prefer token-based authentication integrated with their identity management systems (like Entra ID). Supporting this directly meets a common customer requirement and aligns with Zero Trust principles.
3.  **Best Practices:** We aim to promote cloud security best practices, and leveraging managed identities or user/service principal tokens fits this model better than relying solely on API keys.

While API keys are supported for flexibility, the design encourages the more secure token-based approach as the primary method due to our customer's secure defaults (and above reasons).

Let me know if unsetting the `OPENAI_API_KEY` variable and verifying the Azure-specific ones resolves the error you were seeing with `az login`.

thegovind committed:
fix default AOAI model to o4-mini

thegovind committed:
Merge remote-tracking branch 'origin/gk/azure-openai' into gk/azure-openai

ishowshao commented:
npx codex           

Missing OpenAI API key.
I have logged into Azure by `az login`

theabhinavdas commented:
@ishowshao that won't work since @thegovind's changes haven't been merged and released yet.

@thegovind i've made minimal changes to make this work with an Azure OpenAI API Key (not only do i prefer it, but it seems like a lot of people do). would you like me to fork your repo and issue a PR? might need a little cleanup. 

additionally, i can make it use api key based auth selectively like so:

- if `AZURE_OPENAI_API_KEY` is available, auth with it
- else, try to auth via Entra ID

please let me know your thoughts.

ishowshao commented:
@theabhinavdas i cloned `[thegovind:gk/azure-openai](https://github.com/thegovind/codex/tree/gk/azure-openai)` and installed

ishowshao commented:
@theabhinavdas thanks.  I saw that his PR had not been merged yet, so I cloned his fork locally, installed the dependencies, and used `npx codex` to try to execute his fork. but also get `Missing OpenAI API key.`

theabhinavdas commented:
@ishowshao i'm building from source as described [here](https://github.com/openai/codex?tab=readme-ov-file#installation)

ishowshao commented:
@theabhinavdas 
```
node ./dist/cli.js

Missing OpenAI API key.

Set the environment variable OPENAI_API_KEY and re-run this command.
You can create a key here: https://platform.openai.com/account/api-keys
```

ishowshao commented:
Maybe I didn't do it right, I'll look into it. Thank you anyway @theabhinavdas

theabhinavdas commented:
@ishowshao it's a different (and longer) error in this version as you can see in `./codex-cli/src/cli.tsx` lines 239 onwards. are you sure you're on the right branch?

ishowshao commented:
@theabhinavdas It is indeed the branch reason. I just switched to his branch. Thanks

thegovind commented:
> @ishowshao that won't work since @thegovind's changes haven't been merged and released yet.
> 
> 
> 
> @thegovind i've made minimal changes to make this work with an Azure OpenAI API Key (not only do i prefer it, but it seems like a lot of people do). would you like me to fork your repo and issue a PR? might need a little cleanup. 
> 
> 
> 
> additionally, i can make it use api key based auth selectively like so:
> 
> 
> 
> - if `AZURE_OPENAI_API_KEY` is available, auth with it
> 
> - else, try to auth via Entra ID
> 
> 
> 
> please let me know your thoughts.

Sure, go ahead. The customers I'm working with are in the opposite situation where Entra ID is their only option.

thegovind committed:
add support for AzureOpenai API Key

thegovind committed:
merge main and resolve conflicts

thegovind commented:
> @ishowshao that won't work since @thegovind's changes haven't been merged and released yet.
> 
> @thegovind i've made minimal changes to make this work with an Azure OpenAI API Key (not only do i prefer it, but it seems like a lot of people do). would you like me to fork your repo and issue a PR? might need a little cleanup.
> 
> additionally, i can make it use api key based auth selectively like so:
> 
> * if `AZURE_OPENAI_API_KEY` is available, auth with it
> * else, try to auth via Entra ID
> 
> please let me know your thoughts.

@theabhinavdas  @ishowshao - Found some cycles and added support for `AZURE_OPENAI_API_KEY`. I ran tests and resolved conflicts. Please review and merge (I have several customers waiting on this, and of course the folks following this thread!). cc - @nikodem-wrona

doctorguile commented on codex-cli/src/utils/model-utils.ts:
```diff
@@ -15,6 +25,54 @@ export const RECOMMENDED_MODELS: Array<string> = ["o4-mini", "o3"];
+      const list = await azureClient.models.list();
+      for await (const model of list as AsyncIterable<{ id?: string }>) {
+        if (model && typeof model.id === "string" && (model.id.startsWith('gpt') || model.id.startsWith('o'))) {
+          models.push(model.id);
```
as far as i know, for azure openai, the models.list() is generic and does not reflect your deployments.
so returning this as a list for user to pick is not really useful...

thegovind replied:
I simplified it.

thegovind committed:
simplify model-utils for azure openai to send recommended models and readme updates from main

theabhinavdas reviewed (commented):
LGTM! Thanks, @thegovind 👍 

@tibo-openai Can you please take a look when you get a minute?

scribu commented:
I'm having trouble getting this branch to find my Azure OpenAI deployment (getting 404 Resource not found):

I've defined the following env variables:

```bash
export AZURE_OPENAI_ENDPOINT=https://my-resource-eastus2.cognitiveservices.azure.com/
export AZURE_OPENAI_API_KEY=my-secret-key
```

Details:

```bash
$ npm run build:dev

> @openai/codex@0.1.2504181820 build:dev
> NODE_ENV=development node build.mjs --dev && NODE_OPTIONS=--enable-source-maps node dist/cli-dev.js

╭──────────────────────────────────────────────────────────────╮
│ ● OpenAI Codex (research preview) v0.1.2504181820            │
╰──────────────────────────────────────────────────────────────╯
╭──────────────────────────────────────────────────────────────╮
│ localhost session: a0ea5131c6a14a0ba463a9188c0e5cc0          │
│ ↳ workdir: ~/git/codex/codex-cli                             │
│ ↳ model: o4-mini                                             │
│ ↳ approval: suggest                                          │
╰──────────────────────────────────────────────────────────────╯
user
explain this codebase to me

    system
    ⚠️  OpenAI rejected the request. Error details: Status: 404, Code: 404, Type: unknown, Message: 404 Resource not found. Please verify your
    settings and try again.
```

I've tried different variations for the AZURE_OPENAI_ENDPOINT value:

https://my-resource-eastus2.cognitiveservices.azure.com/openai/deployments/o4-mini
https://my-resource-eastus2.cognitiveservices.azure.com/openai
https://my-resource-eastus2.cognitiveservices.azure.com

None of them seem to work.

theabhinavdas commented:
@scribu I've noticed this error when I didn't have the particular model deployed. In your case, `o4-mini`. Could you check if this might be the issue?

scribu commented:
> @scribu I've noticed this error when I didn't have the particular model deployed. In your case, `o4-mini`. Could you check if this might be the issue?

@theabhinavdas It's definitely deployed. That's where I got the endpoint URL. I can run it through the Azure playground.

thegovind commented:
> @theabhinavdas It is indeed the branch reason. I just switched to his branch. Thanks

@ishowshao - Also upda

> > @scribu I've noticed this error when I didn't have the particular model deployed. In your case, `o4-mini`. Could you check if this might be the issue?
> 
> @theabhinavdas It's definitely deployed. That's where I got the endpoint URL. I can run it through the Azure playground.

@scribu - Could you try it with the URL from [Foundry's Models and Endpoints page](https://ai.azure.com/build/deployments/model). For example, mine is `'https://ai-416build424012653491.openai.azure.com/'`. You can also get this from the Overview blade:
![image](https://github.com/user-attachments/assets/dca72739-cb73-447d-8043-f82df6ae9f14)

Nicowow commented:
> > @scribu I've noticed this error when I didn't have the particular model deployed. In your case, `o4-mini`. Could you check if this might be the issue?
> 
> 
> 
> @theabhinavdas It's definitely deployed. That's where I got the endpoint URL. I can run it through the Azure playground. 

It worked for me when I set the AZURE_OPENAI_API_VERSION to 2025-03-01-preview 👌

APiTJLillo commented:
> > > @scribu I've noticed this error when I didn't have the particular model deployed. In your case, `o4-mini`. Could you check if this might be the issue?
> > 
> > 
> > @theabhinavdas It's definitely deployed. That's where I got the endpoint URL. I can run it through the Azure playground.
> 
> It worked for me when I set the AZURE_OPENAI_API_VERSION to 2025-03-01-preview 👌

Yeah, with Azure OpenAI you have to make sure you grab the right api version. You can check model availability for versions using:
https://<deploymentName>.openai.azure.com/openai/models?api-version=<apiVersion>

scribu commented:
> Could you try it with the URL from [Foundry's Models and Endpoints page](https://ai.azure.com/build/deployments/model). For example, mine is 'https://ai-416build424012653491.openai.azure.com/'. You can also get this from the Overview blade:

@thegovind I've looked through all of the available Azure screens and my URL always ends with `.cognitiveservices.azure.com`

If I manually change it to `.openai.azure.com` I get a 401:

> OpenAI rejected the request. Error details: Status: 401, Code: Unauthorized, Type: unknown, Message: 401 Access denied due to invalid subscription key or wrong API endpoint.

scribu commented:
Working now 👍 (might have been an issue on the Azure side)

dcleres commented:
I have successfully connected Codex to Azure OpenAI using this branch. Thank you very much for the great work! However, when I try to generate code using the tool, I always run into the error message `⚠️  Network error while contacting OpenAI. Please check your connection and try again.` 

This error message is thrown from `codex-cli/src/utils/agent/agent-loop.ts` 

The error happens after a few back-and-forth prompts with Codex. The behavior I have seen so far is that Codex is able to ask some clarifying questions, and then, when generating the actual code, it crashes. Therefore, I assume it is not coming from a connection issue with Azure. 

Has one of you faced the same issue, or in other words, is someone able to generate larger codes based on the code in this PR?

nikhil-pandey commented:
Noticing frequent connection issues — this doesn’t seem to happen with OpenAI. Should we consider adding retries to handle this flakiness with Azure?
![image](https://github.com/user-attachments/assets/95d01233-9aaa-4fed-8200-07b8d6aa5625)

dcleres commented:
To follow-up on my comment above, when I try to reduce the size of the output - by asking for one file at the time, for instance, I was able to reduce the number of network errors while still being able to generate some parts of the code I wanted. However, I agree that handling the network errors from azure in a better way would improve the overall experience. Thank you.

thegovind commented:
Thank you for testing with Azure OpenAI and providing your feedback. I’ve escalated this internally. We are currently experiencing capacity issues due to a spike in demand - both from new users and, more importantly, from migrating existing 4o token consumers to o4-mini or o3. As we increase capacity, you should see fewer timeouts and less general degradation.

doggy8088 commented:
@thegovind Have you implement retry mechanism in this pull request?

abazabaaa commented:
Is there something stopping this from being merged? Not sure how this process works but I also have several people waiting on this.

thegovind commented:
> Is there something stopping this from being merged? Not sure how this process works but I also have several people waiting on this.

I’ve received all the necessary approvals (@theabhinavdas - could you also officially approve the PR as indicated [here](https://github.com/openai/codex/pull/92#pullrequestreview-2780823094)?), and several of you have tested and verified it. At this point, it’s really up to OpenAI, and specifically @tibo-openai.

arynaq commented:
Can we get this in? OpenAI org verification wants my password, no thank you, I will run this via Azure

cevatkerim commented:
/compact didn't work for me I had to make some changes. Can I contribute somehow? @thegovind

thegovind commented:
> /compact didn't work for me I had to make some changes. Can I contribute somehow? @thegovind 

Yes, please go ahead-thank you. As for conflicts, I’ve resolved merge conflicts countless times now. I’ll do it one last time once I know that OpenAI and @tibo-openai intend to support Azure OpenAI. Frankly, it’s starting to look like they don’t, which goes against the spirit of open source.

bsormagec commented:
> > /compact didn't work for me I had to make some changes. Can I contribute somehow? @thegovind
> 
> Yes, please go ahead-thank you. As for conflicts, I’ve resolved merge conflicts countless times now. I’ll do it one last time once I know that OpenAI and @tibo-openai intend to support Azure OpenAI. Frankly, it’s starting to look like they don’t, which goes against the spirit of open source.

Create a new fork similar to open-codex if they're unresponsive; there's no need to handle it that way. :)

heymaaz commented:
> Would it be possible to build this in a way that doesn't depend on "@azure/identity"?

> Yes, please go ahead-thank you. As for conflicts, I’ve resolved merge conflicts countless times now. I’ll do it one last time once I know that OpenAI and @tibo-openai intend to support Azure OpenAI. Frankly, it’s starting to look like they don’t, which goes against the spirit of open source.

I think that they don't want to add the azure/identity dependency @thegovind

thegovind commented:
> > Would it be possible to build this in a way that doesn't depend on "@azure/identity"?
> 
> > Yes, please go ahead-thank you. As for conflicts, I’ve resolved merge conflicts countless times now. I’ll do it one last time once I know that OpenAI and @tibo-openai intend to support Azure OpenAI. Frankly, it’s starting to look like they don’t, which goes against the spirit of open source.
> 
> I think that they don't want to add the azure/identity dependency @thegovind
@heymaaz @bsormagec @tibo-openai 
This approach is not consistent with OpenAI’s official language SDKs, whether [in Python](https://github.com/openai/openai-python/blob/main/pyproject.toml#L66) or in the [recommended Node implementation](https://github.com/openai/openai-node/blob/master/azure.md). Many of the companies I support prohibit static API keys for security reasons and instead require token-based authentication and authorization, which the official SDKs address through the azure-identity library. I do not intend to launch another fork such as “azure-codex” - the goal is a positive-sum outcome that supports both OpenAI and Azure OpenAI so GPT models remain relevant for SWE agents. I trust OpenAI will act in that same spirit of partnership.

jslitzkerttcu commented:
@tibo-openai can we get a response on this?  many of us are waiting for this to be implemented so that we can use the official release and not create a separate fork.

seanabreau commented:
> @tibo-openai can we get a response on this? many of us are waiting for this to be implemented so that we can use the official release and not create a separate fork.

In the [openai-python library](https://github.com/openai/openai-python/blob/main/pyproject.toml#L66), azure-identity is currently listed only as a dev dependency.

Could an alternative approach work here to avoid adding it as a mandatory dependency?

For instance:
  1. Modify the client initialization to accept an Azure AD token string or a token provider function directly.
  2. Document how users can leverage azure-identity (or other methods) in their own code to generate the necessary tokens to pass to the client.

This approach seems like it could provide the needed Azure AD authentication flexibility without imposing the dependency on all users of the library. Would you be open to this solution?

thegovind commented:
> > @tibo-openai can we get a response on this? many of us are waiting for this to be implemented so that we can use the official release and not create a separate fork.
> 
> In the [openai-python library](https://github.com/openai/openai-python/blob/main/pyproject.toml#L66), azure-identity is currently listed only as a dev dependency.
> 
> Could an alternative approach work here to avoid adding it as a mandatory dependency?
> 
> For instance:
> 
> 1. Modify the client initialization to accept an Azure AD token string or a token provider function directly.
> 2. Document how users can leverage azure-identity (or other methods) in their own code to generate the necessary tokens to pass to the client.
> 
> This approach seems like it could provide the needed Azure AD authentication flexibility without imposing the dependency on all users of the library. Would you be open to a solution like this?

@seanabreau - I’m happy to refactor, but first I need clarity on one point: will OpenAI and @tibo-openai support Azure OpenAI as a target for Codex?

Code generation with Codex and GitHub Copilot was our first proof of PMF for transformer models, well before ChatGPT captured the spotlight. Building on that collaborative ethos, now extended to open-source contributions, I trust OpenAI will guide Codex so it remains valuable to developers everywhere, including those on Azure OpenAI.

Frankly, with Claude and Gemini gaining ground, our relevance and usage are no longer where they once were with Dev audience. Ensuring Codex works on Azure OpenAI would be a decisive step toward regaining that momentum.

pablospe commented:
> 2025-04-01-preview inference API in our [data plane](https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview) (current ETA is 4/22)

One question. It seems a bit of delay for this. Any idea why? It seems they are going directly to 2025-05-01: https://github.com/Azure/azure-rest-api-specs/pull/34051?

Nepomuceno commented:
Do we know if OpenAI will give any response about this support ? Is there a way to get it working today using the API key method or not even that work ?

thegovind commented:
> You can open the fork that was made for this pull request and download/run through there. I was able to successfully do it, I can't remember how. I think Manus told me how.

Yes, [my fork](https://github.com/thegovind/codex/tree/gk/azure-openai) will remain available for use with Azure OpenAI. Instructions are [above](https://github.com/openai/codex/pull/92#issue-3000601023). You'll need to set `AZURE_OPENAI_ENDPOINT` and authenticate using either `az login` (Entra ID - token based) or Azure OpenAI API Keys. Provision/retrieve these credentials via [Azure AI Foundry](https://ai.azure.com/)

Nepomuceno commented:
My question was more until this is not merged is there some environment variables that I can set tp be able to use it with azureopenai and an api key from main ?

pablospe commented:
> Do we know if OpenAI will give any response about this support ? Is there a way to get it working today using the API key method or not even that work ?

You might want to try using the litellm proxy and set the `OPENAI_BASE_URL` as a workaround, rather than waiting for this PR to be merged. Here’s an example:  
https://github.com/openai/codex/issues/26#issuecomment-2816595991

While we’ll likely need to be patient for the PR to be merged (it currently has conflicts with the `main` branch), using the litellm proxy is straightforward and I would say generally recommended. It allows you to support a wide range of models without having to update every piece of software to use Azure OpenAI—just point your applications to the proxy and you’re good to go.

chriswl commented:
With the changes on this branch, the AOAI settings are not applied when asking to explain the current command (press `x` in chat when a command needs to be approved). This uses a vanilla OpenAI instance: [code](https://github.com/openai/codex/blob/237f8a11e11fdcc793a09e787e48215676d9b95b/codex-cli/src/components/chat/terminal-chat.tsx#L81)

chriswl commented:
With the changes on this branch, the AOAI settings are not applied when asking to explain the current command (press `x` in chat when a command needs to be approved). This uses a vanilla OpenAI instance: [code](https://github.com/openai/codex/blob/237f8a11e11fdcc793a09e787e48215676d9b95b/codex-cli/src/components/chat/terminal-chat.tsx#L81)

chriswl commented:
With the changes on this branch, the AOAI settings are not applied when asking to explain the current command (press `x` in chat when a command needs to be approved). This uses a vanilla OpenAI instance: [code](https://github.com/openai/codex/blob/237f8a11e11fdcc793a09e787e48215676d9b95b/codex-cli/src/components/chat/terminal-chat.tsx#L81)

CetinSert commented:
Everyone! Let's get this rolling!

fouad-openai commented:
hey @thegovind thanks for this contribution! we landed a PR to add support for multiple providers a ~week ago and would be great if you could use the latest as starting point, here's [providers.ts](https://github.com/openai/codex/blob/main/codex-cli/src/utils/providers.ts) as a reference point

thegovind commented:
> hey @thegovind thanks for this contribution! we landed a PR to add support for multiple providers a ~week ago and would be great if you could use the latest as starting point, here's [providers.ts](https://github.com/openai/codex/blob/main/codex-cli/src/utils/providers.ts) as a reference point

Thank you, @fouad-openai! Working on it now.

thegovind commented:
> hey @thegovind thanks for this contribution! we landed a PR to add support for multiple providers a ~week ago and would be great if you could use the latest as starting point, here's [providers.ts](https://github.com/openai/codex/blob/main/codex-cli/src/utils/providers.ts) as a reference point

Hi @fouad-openai @tibo-openai II'm starting a separate PR (please leave this one open for now) - https://github.com/openai/codex/pull/769 - as the provider approach is quite different from what Codex was released with.

Now, I have 7 tests in main that are failing. I'm not sure if others are seeing this as well (this is on Windows, but I tested in both bash and pwsh). It would be good to resolve this first.

I've tested my changes with Azure OpenAI locally, and they are working fine.

![image](https://github.com/user-attachments/assets/24a2301c-4944-4a3f-b5dc-320dc69a73cd)

assinchu commented:
<img width="684" alt="image" src="https://github.com/user-attachments/assets/718e5298-5095-48e0-b303-4e41a72f7d93" />
 it just hangs, any idea? how to run codex in debug mode

fouad-openai commented:
closing now that #769 merged
